src/bm_sim/P4Objects.cpp

```diff
@@ -648,6 +648,8 @@ P4Objects::init_objects(std::istream *is,
         return 1;
       }
 
+      primitive->_set_p4objects(this);
```

Comment:
this implementation is not correct. Currently a single action primitive instance is created, but if you need a reference to p4objects, then you need to have one instance per p4objects. You can look at what is done for externs, where registering an extern type actually registers a lambda factory function. Unlike for externs though, you need to instantiate only **one** action primitive per p4objects (not a new instance every time the primitive is called in an action!!!).
I suggest that you start by adding a unit test to test_actions which exposes the issue I am trying to describe here: register one primitive action that leverages the new feature you are adding, create 2 different `P4Objects` instances using that primitive inside an action's body, execute the primitives. For example if your primitive action looks up a field list in p4objects, you need to check that both primitives end up with a different field list instance, each in the appropriate p4object for the primitive (but my guess is that with the implementation here you are going to end up with the same).